### ISSUE TITLE
Firefox review feedback

### ISSUE THREAD

**Issue by selfagency** [bs-detector/bs-detector#170]
Good news! We're approved in the Firefox Add-on directory.

Here's what they want to see fixed in the next release:

1) expandLinks asynchronously expands the links. So the sendResponse call at line 146-149 in background.js will not contain all links. I suggest to wait until all request finish before calling sendResponse. You can asynchronously call sendResponse by returning true inside the onMessage listener. 

2) bs-detector.js, line 205 - 212 is not working at best, and allows remote code execution at worst. 
`this.shorts` is an array set to the value of the "shorteners" key sent by background.js, which in turn is *currently* set to a list of domain names. The following snippet from line 207, 
'a[href*="' + $(this) + '"]'; 
will *currently* evaluate to 'a[href*='[object Object]']', and usually result in an empty list. 
You probably want to change the getLinks function to: 

var shorts = this.shorts; 
var selectors = $('a[href]').filter(function(index, a) { 
return shorts.some(function(shortener) { 
return a.hostname.endsWith(shortener); 
}); 
}).each(function(index, a) { 
bsd.toExpand.push(a.href); 
}); 

This suggestion addresses the following issues: 
(1) $(this).attr('href') can return a non-absolute URL (when the href attribute is not an absolute URL). 
(2) The original check (after replacing $(this) with this at line 207 would also match URLs that contain the substring, e.g. example.com/bit.do. The proposed code only returns the URL if it is a domain or subdomain in the shorts list. 
(3) The original code can insert the same link multiple times, if its URL contains multiple substrings from the shorts array. 
(the current method can still contain duplicates, if you want to eliminate those too, use jQuery.unique to filter duplicates) 

3) line 344 in bs-detector.js puts text in a HTML context. This may cause security issues in the future, please change the line to
$('.bs-alert').append($('<span>').text(this.warnMessage)); 
(similarly for lines 481 and 483 in the flagPost function) 

4) The add-on makes use of external web services. This should be disclosed in the add-on description.

**Comment by Bastlynn** [comment]
I've got a branch open on my local and starting work on these.

**Comment by Bastlynn** [comment]
PR is up - does not address #4 since that's not a code thing.

**Comment by thinkcontext** [comment]
Regarding expandLinks, I'd suggest making async work rather than waiting for all links to resolve and sending in a batch (synchronously).  Say there are 10 shortened links on a page and when unshortening the 10th one stalls with 1/2 second delay.  When using asynchronous style the UI for 9 links can be displayed immediately, the 10th will lag (if its even noticed).  With synchronous style the display of all links will lag and will definitely be noticed.

**Comment by tomlutzenberger** [comment]
I agree with @thinkcontext. This might lead to delays and therefore bad UX.

**Comment by Bastlynn** [comment]
Done! Now using a callback on success to return the link instead of waiting for all processing with a defer.
